drivers/pwm/pwm_npcx.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_pwm
+
+#include <assert.h>
+#include <drivers/pwm.h>
+#include <dt-bindings/clock/npcx_clock.h>
+#include <drivers/clock_control.h>
+#include <kernel.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_npcx, LOG_LEVEL_ERR);
+
+/* 16-bit period cycles/prescaler in NPCX PWM modules */
+#define NPCX_PWM_MAX_PRESCALER      (1UL << (16))
+#define NPCX_PWM_MAX_PERIOD_CYCLES  (1UL << (16))
+
+/* PWM clock sources */
+#define NPCX_PWM_CLOCK_APB2_LFCLK   0
+#define NPCX_PWM_CLOCK_FX           1
+#define NPCX_PWM_CLOCK_FR           2
+#define NPCX_PWM_CLOCK_RESERVED     3
+
+/* PWM heart-beat mode selection */
+#define NPCX_PWM_HBM_NORMAL         0
+#define NPCX_PWM_HBM_25             1
+#define NPCX_PWM_HBM_50             2
+#define NPCX_PWM_HBM_100            3
+
+/* Device config */
+struct pwm_npcx_config {
+	/* pwm controller base address */
+	uintptr_t base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const uint8_t   alts_size;
+	const struct npcx_alt *alts_list;
+};
+
+/* Driver data */
+struct pwm_npcx_data {
+	/* PWM cycles per second */
+	uint32_t cycles_per_sec;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct pwm_npcx_config *)(dev)->config)
+
+#define DRV_DATA(dev) ((struct pwm_npcx_data *)(dev)->data)
+
+#define HAL_INSTANCE(dev) (struct pwm_reg *)(DRV_CONFIG(dev)->base)
+
+/* PWM local functions */
+static void pwm_npcx_configure(const struct device *dev, int clk_bus)
+{
+	struct pwm_reg *inst = HAL_INSTANCE(dev);
```

Comment:
nit: here and all functions we can add another const modifier

```suggestion
	struct pwm_reg *const inst = HAL_INSTANCE(dev);
```

Reply:
Done.

Reply:
I was thinking that you would also update the other variables pointers as well to have the `const` modifier, like `config`, `data`, and `clk_dev` in other functions.

If you have to make other changes in this file, you can make these updates. If there are no other required file updates, then you can leave it the way it is.

Reply:
Actually, I think it's a good discipline for this driver and following npcx drivers.  And I will update it to previous drivers in a new commit for consistency.

Reply:
This is a coding style decision that we really ought to get consensus on before making wholesale changes, because if we're going to do it we should do it consistently, and following a rule.  Historically only a few drivers qualified the pointers that they copied out of immutable structures to local (const) variables.  Should every variable definition that the developer/maintainer believes will not be changed be marked const?

Making local variables `const` does clarify intent, but is there a clear risk to standardizing on unqualified local "variables"?

It's not a particularly big deal (and I do this frequently myself in C++), but we use Coccinelle to automate code transformations, and when those transformations involve adding or removing `const` qualifiers the presence of extra `const` qualifiers makes it difficult to get the transformation rules correct.

I've also seen this in API function declarations to indicate the parameter variable will not be changed, which I do object to.

Reply:
These are new files we are adding, so I think it is okay to suggest and make the changes here. I think within a file, we should definitely be self consistent; that is why I recommend the other variables in this file to also add the `const` modifier.

I like the idea of including this in Zephyrs style (or proposing it at least). I do wonder if there is a coverity option we can turn on somewhere.

I think use of `const` should be used when it is possible as it can prevent accidental coding mistakes as well as self-document the variables purpose better.